python/pylibvw.cc

```diff
@@ -540,32 +540,95 @@ std::string varray_char_to_string(v_array<char>& a)
 
 std::string my_get_tag(example_ptr ec) { return varray_char_to_string(ec->tag); }
 
-uint32_t ex_num_namespaces(example_ptr ec) { return (uint32_t)ec->indices.size(); }
+uint32_t ex_num_namespaces(example_ptr ec)
+{
+  uint32_t ret = 0;
+  for (auto& group_list : ec->feature_space) 
+  {
+    ret += group_list.size();
+  }
+   return ret;
+}
 
-unsigned char ex_namespace(example_ptr ec, uint32_t ns) { return ec->indices[ns]; }
+unsigned char ex_namespace(example_ptr ec, uint32_t distance) 
+{
+  uint32_t curr = 0;
+  for (auto& group_list : ec->feature_space) 
+  {
+    if (curr + group_list.size() > distance)
+    {
+      auto it = group_list.begin();
+      std::advance(it, distance - curr);
+      return it->index;
+    }
+    else
+    {
+      curr += group_list.size();
+    }
+  }
+   return -1;
+}
 
-uint32_t ex_num_features(example_ptr ec, unsigned char ns) { return (uint32_t)ec->feature_space[ns].size(); }
+uint32_t ex_num_features(example_ptr ec, const char* ns) {
+  auto& fs = ec->feature_space;
+  uint64_t hash = hashstring(ns, strlen(ns), 0);
+  features* fg = fs.get_or_null(ns[0], hash);
+  if (fg == nullptr)
+  {
+    return fs.get_list(ns[0]).front().feats.size();
```

Comment:
If this is null shouldnt you throw/return a not found error?

EDIT: Oh I see is this the non-hash version? That's really confusing I would really like to see it very explicit if the character or the hash is being used to look up a feature group